feat: rewrite microphone-stream in TypeScript

trivikr opened the pull request:
Fixes: https://github.com/microphone-stream/microphone-stream/issues/41
Fixes: https://github.com/microphone-stream/microphone-stream/issues/46

~~This draft PR is posted to get initial feedback.~~ PR is ready for review! 🎉 

This code should be released as ~~v6.0.0~~ v6.0.0-beta.1 as it contains BREAKING CHANGE: MicrophoneStream no longer accepts legacy stream as first parameter. It needs to be passed in options.

The single options parameter was introduced in `v4.0.1` in https://github.com/microphone-stream/microphone-stream/commit/ef410f90933b58de0c50a57fa7672099cc7769ae

trivikr committed:
chore: rename src file from js to ts

trivikr committed:
chore: update tooling to transpile TypeScript code

trivikr committed:
chore: tsconfig alwaysStrict=true

trivikr committed:
chore: create MicrophoneStreamOptions interface

trivikr committed:
feat: convert microphone-stream to TypeScript

trivikr committed:
chore: import Readable

trivikr committed:
chore: remove unused util

trivikr committed:
chore: import bufferFrom

trivikr committed:
chore: annotate types from JSDoc

nfriedly commented:
Awesome!

A couple of thoughts:

* We might want to keep the existing JS tests as-is to ensure nothing was lost in the TS conversion. Maybe add new TS tests along side them.
* I'd prefer to check in the generated JS, similar to what we're currently doing with the generated types. This allows for people to install directly from github. (I know github releases are a thing, but they don't seem to be as widely used as installing form the repo.)

trivikr commented:
> We might want to keep the existing JS tests as-is to ensure nothing was lost in the TS conversion

I plan to convert existing tests to TypeScript, will prioritize it later (may be next weekend).

> I'd prefer to check in the generated JS, similar to what we're currently doing with the generated types. This allows for people to install directly from github.

Should I remove `dist` from `.gitignore` and check-in `dist/microphone-stream.js`?
If customers plan to ue `<script>` tag to include the output JavaScript directly, I'll reintroduce JSDoc in TypeScript.

trivikr committed:
chore(deps-dev): add @typescript-eslint/*

trivikr committed:
fix: run eslint on src/**/*.ts

trivikr committed:
fix: eslint errors

trivikr committed:
chore: only make context and audioInput public

trivikr committed:
chore: attempt to move tests to TypeScript

trivikr committed:
chore: move spec.ts in src folder

trivikr committed:
fix: karma unable to determine file type fro the file extension

SO: https://stackoverflow.com/a/64921199/477683

trivikr committed:
fix: executed 0 of 0 error from karma-typescript

Reason: karma does not host typescript files by default.
https://github.com/karma-runner/karma/issues/753#issuecomment-398852991

trivikr committed:
fix: include spec.ts files in karma.conf.js

trivikr committed:
fix: cannot read property 'createScriptProcessor' of undefined

trivikr committed:
fix: eslint errors in spec.ts

trivikr committed:
fix: pass stream as part of options

trivikr commented:
Test files are moved to TypeScript, and most errors are fixed.

The error with test appears to be either with the karma-typescript setup or the package:
```console
Chrome 88.0.4324.150 (Mac OS 10.15.7) MicrophoneStream should capture audio and emit AudioBuffers when in object mode FAILED
	TypeError: The "chunk" argument must be one of type string, Buffer, or Uint8Array. Received type object
	    at new NodeError (/var/folders/1v/nghjtx0s3hg_s7xz_73cjjg47fny27/T/karma-typescript-bundle--78206-TteUeoEbTqEP-.js:876:20)
	    at chunkInvalid (/var/folders/1v/nghjtx0s3hg_s7xz_73cjjg47fny27/T/karma-typescript-bundle--78206-TteUeoEbTqEP-.js:6463:10)
	    at readableAddChunk (/var/folders/1v/nghjtx0s3hg_s7xz_73cjjg47fny27/T/karma-typescript-bundle--78206-TteUeoEbTqEP-.js:6408:31)
	    at MicrophoneStream.Readable.push (/var/folders/1v/nghjtx0s3hg_s7xz_73cjjg47fny27/T/karma-typescript-bundle--78206-TteUeoEbTqEP-.js:6391:10)
	    at MicrophoneStream.<anonymous> (/var/folders/1v/nghjtx0s3hg_s7xz_73cjjg47fny27/T/karma-typescript-bundle--78206-TteUeoEbTqEP-.js:296:22)
```

trivikr committed:
fix: use named import of Readable to extend

trivikr committed:
chore: target ES5 in TSConfig

trivikr committed:
fix: pass objectMode to super call

trivikr committed:
chore: use let/const in test instead of var

trivikr committed:
chore(readme): use const in examples

trivikr committed:
fix: remove redundant tsconfig from karama.conf.js

Refs: https://github.com/monounity/karma-typescript/issues/127#issuecomment-312411684

jackiosport commented:
Looks good! 

Perhaps not related to this MR, but you could use arrow functions to avoid the use of `bind`

trivikr committed:
fix: use arrow functions to avoid bind

trivikr commented:
> Looks good!

Thank you!

> Perhaps not related to this MR, but you could use arrow functions to avoid the use of `bind`

Usage of bind removed in [`657424c` (#48)](https://github.com/microphone-stream/microphone-stream/pull/48/commits/657424c713c22c53ee722c48cb69b24d1803a6f5) :+1:

trivikr commented:
> I'd prefer to check in the generated JS, similar to what we're currently doing with the generated types.

@nfriedly I'd prefer to release this change separately as `v6.0.0-beta.2` if you're ok.
We can make other updates in subsequent beta and rc releases (if needed), waiting for feedback from community before releasing `v6.0.0`

nfriedly commented on src/microphone-stream.ts:
```diff
@@ -0,0 +1,208 @@
+  /**
+   * no-op, (flow-control doesn't really work on live audio)
+   */
+  public _read(/* bytes */): void {
```
I don't think this needs to be `public`. It should only be called by the `Readable` class that this extends, so that's either `private` or `protected`.

trivikr replied:
The `_read()` function is neither public nor private in @types/readable-stream: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/2fee823196d4cc536fbc400086f1ccc582ebd128/types/readable-stream/index.d.ts#L87

TypeScript class members are public by default https://www.typescriptlang.org/docs/handbook/classes.html#public-by-default

nfriedly commented:
> Should I remove dist from .gitignore and check-in dist/microphone-stream.js?

Yes, that's what I was thinking.

> If customers plan to use <script> tag to include the output JavaScript directly,

That's one potential use, but also when doing `npm install https://github.com/microphone-stream/microphone-stream#branch` (or tag or commit hash) - that's something that I find myself doing occasionally to get bugfixes and whatnot that haven't yet been published to npm, or sometimes not even merged into the main repo, and it always annoys me when it doesn't work because they didn't check in the compiled js.

> I'll reintroduce JSDoc in TypeScript.

I don't think that's important. I wouldn't fight you on it, but I think I'd prefer to just use TypeScript to define the API rather than trying to keep TS and JSDoc in sync.

---

> > I'd prefer to check in the generated JS, similar to what we're currently doing with the generated types.
>
> @nfriedly I'd prefer to release this change separately as v6.0.0-beta.2 if you're ok.
We can make other updates in subsequent beta and rc releases (if needed), waiting for feedback from community before releasing v6.0.0

What's your reasoning here? I was thinking this would be a 1-line change, just removing `dist` from .gitignore.

trivikr committed:
chore: check-in dist folder

trivikr commented:
> > > I'd prefer to check in the generated JS, similar to what we're currently doing with the generated types.
> > 
> > 
> > @nfriedly I'd prefer to release this change separately as v6.0.0-beta.2 if you're ok.
> > We can make other updates in subsequent beta and rc releases (if needed), waiting for feedback from community before releasing v6.0.0
> 
> What's your reasoning here?

I wanted to explore GitHub Release is a separate PR.
But never mind, I've added dist folder in [`3e3a3f1` (#48)](https://github.com/microphone-stream/microphone-stream/pull/48/commits/3e3a3f1274cfc79e5ecc69f1aa66ced1ec9d039c)

trivikr commented:
Is this PR good to merge @nfriedly ?

nfriedly reviewed (approved):
Yea, I think this looks good!
